Implement ComplexDataVector

moxcodes opened the pull request:
## Proposed changes

Add the class `ComplexDataVector`, along with all attendant specializations
appropriate for a spectre Vector storing complex variables.

- defined a specialization of
`VectorImpl<std::complex<double>,ComplexDataVector>`
- ComplexDataVectors have arithmetic operations defined for DataVectors and
doubles just as `std::complex<double>`s have arithmetic operations with
doubles
- some math functions do not work on ComplexDataVectors, but imag, abs, and
real return DataVectors (real values)
- tests take heavily from `Test_DataVector.cpp`, but include tests of math
operations between ComplexDataVector and real types

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

- [x] Depends on #1247

nilsdeppe commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,246 @@
+ComplexDataVector conj(const ComplexDataVector& t) noexcept;
+
+namespace blaze {
+  DATAVECTOR_BLAZE_TRAIT_SPEC(ComplexDataVector)
```
Please figure out how to get your editor to do the right formatting.

moxcodes replied:
Sorry. Done (both formatting this and setting emacs).

nilsdeppe commented on tests/Unit/DataStructures/Test_ComplexDataVector.cpp:
```diff
@@ -0,0 +1,749 @@
+// Nontrivial changes primarily in math tests: ComplexDataVectors are
+// mathematically different but not structurally different
+
+SPECTRE_TEST_CASE("Unit.DataStructures.ComplexDataVector",
```
Is it really not possible to share some/most of these tests with DataVector? This also seems like a lot of copy-paste

moxcodes replied:
I agree - it is a ton of copy-paste. As a first start, I kept them silo-ed, but all aside from math operations could be directly duplicated, running once for each type. The math operations would need significantly more care (because reals and complex actually act differently mathematically), but I could work to unify them it if it's significantly preferable.

moxcodes commented:
This one is updated as well, but depends on DataVectorImpl, which depends on other PRs too.

nilsdeppe commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,182 @@
+template <int N>
+struct pow<ComplexDataVector, N, Requires<(N < 0)>> {
+  static_assert(N > 0,
+                "Cannot use pow on DataVector structures with a negative "
```
`DataVector` -> `ComplexDataVector` and I guess the formatting of the string (the awkward line break on the second line) could be fixed :)

nilsdeppe commented on src/DataStructures/DataVector.hpp:
```diff
@@ -51,7 +51,8 @@ class DataVector : public VectorImpl<double, DataVector> {
 
 namespace blaze {
-VECTOR_BLAZE_TRAIT_SPEC(DataVector);
+VECTOR_BLAZE_TRAIT_SPEC_ARITHMETIC(DataVector);
```
could you factor the change with the new macro into a separate commit please?

nilsdeppe commented on tests/Unit/DataStructures/DataBox/Test_DataBox.cpp:
```diff
@@ -27,12 +27,11 @@
 #include "Utilities/TypeTraits.hpp"
 #include "tests/Unit/TestHelpers.hpp"
 
+// IWYU pragma: no_forward_declare Variables
```
This seems random and unrelated. I assume it'll disappear since I think the `Variables` generalizations were merged already.

nilsdeppe commented on tests/Unit/DataStructures/Test_ComplexDataVector.cpp:
```diff
@@ -0,0 +1,159 @@
+
+class DataVector;
+
+SPECTRE_TEST_CASE("Unit.DataStructures.ComplexDataVector",
```
The same comment as on `ModalVector` PR about combining with the `Math` test and switching from `SECTION` to `INFO`.

nilsdeppe commented on tests/Unit/DataStructures/Test_ComplexDataVector.cpp:
```diff
@@ -0,0 +1,159 @@
+}
+
+// [[OutputRegex, Must copy into same size]]
+[[noreturn]] SPECTRE_TEST_CASE(
```
I think there is also another error test missing in this one :)

nilsdeppe reviewed (commented):
Also figured I'd look at this one since we're quite close with the backend changes. Again just a few minor comments. It looks like this will need a bit more work during the rebase but that should also clean the commit up quite a bit :)

moxcodes committed:
Adjust utilities for complex<double> support

- changed testing utilities
- added case for MakeWithValuesImpl
- adjusted generic math tests for wider support of types

moxcodes committed:
Utilities for Real, Conj, Imag

- Blaze has its real, imag, conj in the global namespace, so for easy overload
  use between vector types and std::complex types, the std::real, std::imag and
  std::conj are moved to global namespace as well.
- add dereference wrappers for Real, Conj, and Imag

moxcodes committed:
Split VectorImpl Blaze trait specialization macro

... into two new macros, one for arithmetic traits, and one for map traits.

moxcodes committed:
Implement ComplexDataVector

Add the class `ComplexDataVector`, along with all attendant specializations
appropriate for a derived class of VectorImpl for complex values.

- defined a specialization of
`VectorImpl<std::complex<double>,ComplexDataVector>`
- tests based on VectorImpl testing framework

moxcodes committed:
Add ComplexDataVector support

... to Tensor and Variables tests.

moxcodes commented:
Rebased and made several adjustments from the VectorImpl changes. I think I've also addressed all of the current comments on the code.
Cheers!

kidder commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -496,15 +520,19 @@ void test_function_on_vector_operands(
   MAKE_GENERATOR(generator);
   UniformCustomDistribution<size_t> size_distribution{2, 5};
-  const DataVector used_for_size{size_distribution(generator)};
+  //  const DataVector used_for_size{size_distribution(generator)};
```
remove

kidder commented on tests/Unit/TestingFramework.hpp:
```diff
@@ -138,6 +138,16 @@ struct check_iterable_approx {
   }
 };
 
+template <typename T>
```
test this in Test_TestHelpers?

kidder commented on src/Utilities/MakeWithValue.hpp:
```diff
@@ -53,6 +53,14 @@ struct MakeWithValueImpl<double, T> {
   }
 };
 
+template <typename T>
```
add a test for this (in Utilities/Test_MakeWithValue)?

kidder commented on tests/Unit/DataStructures/VectorImplTestHelper.hpp:
```diff
@@ -446,7 +446,9 @@ struct VectorOrArraySize {
   // vector version
   template <typename T,
-            Requires<cpp17::is_arithmetic_v<typename T::ElementType>> = nullptr>
+            Requires<cpp17::is_arithmetic_v<typename T::ElementType> or
```
are these changes tested anywhere?

moxcodes replied:
I've added a test for the custom size and element functions in `Test_VectorImplTestHelper.cpp`. For the order of the commits, the fixup to the utilities commit includes only tests of `DataVector`, and the additional tests for `ComplexDataVector` are added in the fixup commit for implementing `ComplexDataVector`.

kidder commented on src/DataStructures/VectorImpl.hpp:
```diff
@@ -435,29 +435,43 @@ std::ostream& operator<<(std::ostream& os,
+ * by specializing Blaze's type traits that are used for handling return type
+ * deduction.
+ *
+ * \details Type declarations here are suitable for the original DataVector, but
```
same change as above? (contiguous data, e.g. DataVector)

kidder commented on src/DataStructures/VectorImpl.hpp:
```diff
@@ -435,29 +435,43 @@ std::ostream& operator<<(std::ostream& os,
+ * this macro might need to be tweaked for other types of data, for instance
+ * Fourier coefficients
  *
+ * \param VECTOR_TYPE The vector type, which for the arithmetic operations is
```
`Map` operations

kidder commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,179 @@
+// Distributed under the MIT License.
```
I don't know what *Vector is in the commit message

kidder commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,179 @@
+ * available types.
+ *
+ * ComplexDataVectors support a variety of mathematical operations that are
+ * applicable to nodal coefficients. In addition to common arithmetic operations
```
why are nodal coefficients mentioned?

kidder commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,179 @@
+ * ComplexDataVectors support a variety of mathematical operations that are
+ * applicable to nodal coefficients. In addition to common arithmetic operations
+ * such as elementwise addition, subtraction, multiplication and division,
+ * elementwise operations on blaze vectors of doubles are supported. See
```
I don't understand what `elementwise operations on blaze vectors of doubles` are

kidder commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,179 @@
+  }
+};
+
+template <int N>
```
is this tested? It looks like the Requires and static_assert are mutually exclusive 😕

moxcodes replied:
Hmmm, I agree this is fishy. I've just duplicated this directly from DataVector under the assumption that there was a good (likely blaze-related) reason for these to be here. Now that you point it out, I think I'd prefer to eliminate these and just trust the Blaze pow, unless there's a compelling reason to keep them. @nilsdeppe can you shed some light on what these power functions are used for?

nilsdeppe replied:
Yea, these should be tested. No, you really don't want to defer to a runtime pow, it's horribly slow. I can't remember the exact results but when I compared the computations a year and a half ago at least for small N (less than 10ish or something like that) pow was a factor of several slower than the multiplications. SpEC goes so far as to implement a pow that's runtime but faster if the exponent is an int. We should generally use this version of pow whenever we know the exponent at compile time.

moxcodes replied:
Ah, thanks, that helps. Although, I'm still a bit confused. Given the point made by @kidder, and that the self-contradictory implementation was present before I started working on the DataVectors, it seems any compile-time `pow`s currently used in the code base must simply never use negative powers?

I'll add tests, and I think these utilities can be tidied up a bit in the process.

nilsdeppe replied:
It looks like there's no actual issue with that code. If someone tries to use it then they get the static assert. I'm not sure though that it couldn't be done more elegantly (combined in the general case).

moxcodes replied:
I'd feel pretty strongly that if the assert should be there, the implementation details shouldn't. otherwise, it is confusing which is intended when reading the code. I'll have a try at making it clearer, but either way, I'll also add documentation for how it ends up.

edit: @nilsdeppe, do you know why this function wants to prevent usage with negative exponents? The implementation looks like it handles the trouble, but I don't want to cause subtle bugs that you've already solved in the past. :)

nilsdeppe replied:
So the issue with the implementation is that it allocates memory that doesn't need to be allocated if dividing by a positive exponent. Does that make sense?

moxcodes replied:
Ah, got it. I was concerned that it might be causing extra allocations. Great, thanks! I'll see what I can do.

kidder commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,179 @@
+  using Type = DataVector;
+};
+template <>
+struct UnaryMapTrait<DataVector, blaze::Imag> {
```
what does this do?

moxcodes replied:
It should always return 0, but this should be present for compatibility with the `std::imag`, which returns 0 for floating-point fundamentals.

kidder replied:
okay

moxcodes committed:
handle vector types in ConstantExpressions pow

- remove specializations in DataVector
- add Functional UnaryPow
- add tests for functional and vector use of pow<N>

kidder commented on tests/Unit/DataStructures/Test_DataVector.cpp:
```diff
@@ -28,9 +28,9 @@
-// [[OutputRegex, Must copy into same size]]
-[[noreturn]] SPECTRE_TEST_CASE("Unit.DataStructures.DataVector.RefDiffSize",
-                               "[DataStructures][Unit]") {
+    // [[OutputRegex, Must copy into same size]]
```
what happened here?

fmahebert replied:
`clang-format` often gets error tests wrong in my experience :(

nilsdeppe replied:
Should be able to get around this by using:
```cpp
// clang-format off
// [[OutputRegex, ...]]
SPECTRE_TEST_CASE() {
  // clang-format on
```

kidder reviewed (commented):
one formatting issue, otherwise rebase and squash

moxcodes commented:
squashed and formatting fixed.
Cheers!

kidder commented:
@wthrowe , if you have time can you please look at this (or let us know you don't have time)

wthrowe commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -0,0 +1,142 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class Data.
```
Remove.

wthrowe commented on src/DataStructures/VectorImpl.hpp:
```diff
@@ -435,29 +435,43 @@ std::ostream& operator<<(std::ostream& os,
  * \param VECTOR_TYPE The vector type, which for the arithmetic operations is
  * the type of the operation result (e.g. `DataVector`)
+ */
+#define VECTOR_BLAZE_TRAIT_SPECIALIZE_ARITHMETIC_TRAITS(VECTOR_TYPE) \
```
The documentation for this macro (and the next one) does not appear to be showing up anywhere.

wthrowe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,83 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
+///
+/// \warning For optimization, it is assumed that the `pow<0>` of a vector type
+/// (e.g. `DataVector`) will not be used for initialization or directly indexed,
+/// so `pow<0>` returns simply `1` in an appropriate fundamental type. In the
```
The "appropriate fundamental type" is always a double if I'm reading the code correctly.

wthrowe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,83 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
+/// \tparam N the integer power being raised to in \f$t^N\f$
 /// \param t the value being exponentiated
-/// \return value t^N of type T
+/// \return value \f$t^N\f$ of type T
```
Not broken by this PR, but the return type of this function is often not T.  For T=DataVector it can be double, DataVector, or various expression template types depending on N.

wthrowe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,83 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
+template <typename T>
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& /*t*/, std::integral_constant<int, 0> /*meta*/,
+    cpp17::bool_constant<false> /*meta*/) noexcept {
```
The recursion between the various `pow_impl` functions is not working the way you want, resulting in extra multiplications for N not 0, 1, or -2.  This overload should be removed, ...

wthrowe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,83 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
-/// \endcond
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& t, std::integral_constant<int, 1> /*meta*/,
+    cpp17::bool_constant<false> /*meta*/) noexcept {
```
... this one should take `cpp17::bool_constant<true>` as it's last argument, ...

wthrowe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,83 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
+template <int N, typename T>
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& t, std::integral_constant<int, N> /*meta*/,
+    cpp17::bool_constant<false> /*meta*/) noexcept {
```
... and this one should not do a multiplication and should just call the `-N` version for the denominator.  Finally, the `pow`  function itself should dispatch using `cpp17::bool_constant<(N >= 0)>`.

The reasons for the extra multiplications fell into three groups.  Because the `bool_constant` changed between 1 and 2, the `N >= 2` cases were using 0 instead of 1 as the base case, computing t^1 as `t * 1` instead of `t`.  The `N == -1` case was explicitly computing its denominator as `t * pow<0>(t)`.  The `N <= -3` cases were fine themselves, but called the `N >= 2` cases and so inherited their bug.

moxcodes replied:
I do agree this is a peculiar structure - and I initially implemented it as you suggest, but the execution flow is intentional. After checking performance tests, I found counter-intuitively that just returning t early only really helped in the case of the power being exactly 1 and in larger powers (>4). For more practical powers 2-4, proceeding in the recursion was actually faster. My rough hypothesis was that involving the `return t` function had a cost even in excess of an extra multiplication. If you like, I can dig up the set of tests I performed.

It is true that for larger powers the version that seems fastest is fastest, but I was mostly trying to make sure that this generic form retained the speed of the former version for what seemed like the most often used powers.
edit: the case of -2 is a good point - I don't think I tried that one.

wthrowe replied:
I tried it and got the same result as you.  But it has nothing to do with the recursion (which should all be optimized away anyway).  If I just evaluate `r = a * b`, `r = (a * b) * 1.0`, and `r = a * (b * 1.0)` with `r`, `a`, and `b` all DataVectors of the same size (100 for my test), the two triple products are both *faster* than the double product.  All the expressions are faster if `b` is replaced by `a` to get a squaring expression, but calculating the square by `a * a` is still the slowest.

@nilsdeppe do you have any idea what is going on here?

nilsdeppe replied:
I looked at the generated assembly and am not sure what's going on. Here's a plot I did on my machine:

![powspeedcomparison](https://user-images.githubusercontent.com/13531030/51196328-e7991b80-18bc-11e9-8a97-d77c44760b19.png)
While there is systematic better performance when multiplying by 1, you can see that overall the results are fairly noisy and dependent on exactly what size is being used since that changes how many non-vectorized instructions are necessary to get each grid point. I do think this should be reported upstream to Blaze once we have verified it with their code on the latest release/develop commit.

My conclusion is that you're trying to optimize for a case that we don't know if it matters. Please go back to the old code implementation that is simpler.

moxcodes replied:
The new version's complexity is largely to avoid SFINAE and to grab some extra speed when computing `pow<1>`. If the speed difference for the *1.0 small enough and mysterious enough to ignore, I'll revert to the code that @wthrowe described, which has the intuitive optimizations minimizing number of multiplications.

nilsdeppe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,83 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
-template <typename T, int N>
-struct pow<T, N, Requires<(N < 0)>> {
-  SPECTRE_ALWAYS_INLINE static constexpr decltype(auto) apply(const T& t) {
-    return static_cast<T>(1) / (t * pow<T, -N - 1>::apply(t));
```
I don't understand why you need to do anything more than change this one like to using `get_fundamental_type` instead of `T`.

moxcodes commented:
I think I've fixed all of the comments in the fixups. I've gone ahead and fixed the `pow` as in @wthrowe 's comments, despite a small and mysterious performance impact.

nilsdeppe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,73 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& /*t*/, std::integral_constant<int, 0> /*meta*/,
+    cpp17::bool_constant<true> /*meta*/) noexcept {
+  return 1.0;
```
Shouldn't this do a `static_cast<T>(1)` like was done before, or the one with get_fundamental_type that you added in the negative exponent case?

nilsdeppe commented on src/Utilities/ConstantExpressions.hpp:
```diff
@@ -89,45 +90,73 @@ constexpr uint64_t factorial(const uint64_t n) noexcept {
-/// \endcond
+SPECTRE_ALWAYS_INLINE constexpr decltype(auto) pow_impl(
+    const T& t, std::integral_constant<int, 1> /*meta*/,
+    cpp17::bool_constant<true> /*meta*/) noexcept {
```
it might be useful to replace the name `meta` of the bool constant with `exponent_was_positive` since just a `bool_constant` in this context is not super clear.

nilsdeppe reviewed (commented):
Just two small comments, overall I'm fine with this.

moxcodes commented:
Updated with fixup to last couple of comments.
Cheers!

nilsdeppe reviewed (commented):
Once @wthrowe has looked at it I'm happy for you to squash :)

wthrowe commented:
Looks good.  Squash.

moxcodes commented:
Squashed.
Cheers!
